src/coreComponents/fieldSpecification/FieldSpecificationManager.cpp

```diff
@@ -195,26 +250,65 @@ void FieldSpecificationManager::validateBoundaryConditions( MeshLevel & mesh ) c
       GEOS_THROW( errorMessageBuilder.str(), InputError );
     }
 
-    // if a target set is empty, we issue a warning
-    // ideally we would just stop the simulation, but the SurfaceGenerator relies on this behavior
-    for( auto const & mapEntry : isTargetSetEmpty )
+    if( !setTypesMap.empty() && MpiWrapper::commRank() == 0 )
     {
-      GEOS_LOG_RANK_0_IF( ( mapEntry.second == 1 ), // target set is empty
-                          GEOS_FMT( "\nWarning!\n{}: this FieldSpecification targets (an) empty set(s) within `{}`"
-                                    "\nIf the simulation does not involve the SurfaceGenerator, check the content of the set `{}`. \n",
-                                    fs.getDataContext(), fs.getObjectPath(), mapEntry.first ) );
+      std::ostringstream message;
+      message << GEOS_FMT( "{}: this FieldSpecification targets (an) empty set(s).\n",
+                           fs.getDataContext() );
+
+      for( auto const & [setName, objectTypesPair] : setTypesMap )
+      {
+        forManagerInMeshLevel< FaceManager,
+                               EdgeManager,
+                               NodeManager >( mesh, expectedSetType, setName, setTypesMap );
+
+        string_array capturedTypes;
+        for( auto const & pair : objectTypesPair )
+        {
+          if( pair.first != expectedSetType )
+          {
+            capturedTypes.push_back( EnumStrings< MeshObjectPath::ObjectTypes >::toString( pair.first ) );
+          }
+        }
+
+        if( !capturedTypes.empty())
+        {
+          message << GEOS_FMT( "Set '{}':\n"
+                               "  - Does not capture: {}\n", setName, fs.getObjectPath());
+          message << GEOS_FMT( "  - Instead, captures: {}\n", stringutilities::join( capturedTypes, ", " ));
+        }
+        else
+        {
+          message << GEOS_FMT( "Set '{}' does not capture anything in the mesh ", setName );
+        }
+      }
+      string const cc = "error";
+      Wrapper< string > const & wrapper = fs.getWrapper< string >(
+        FieldSpecificationBase::viewKeyStruct::errorSetModeString());
+      switch( EnumStrings< FieldSpecificationBase::SetErrorMode >::fromString( wrapper.getDefaultValue() ))
+      {
+        case  FieldSpecificationBase::SetErrorMode::silent:
+          break;
+        case  FieldSpecificationBase::SetErrorMode::error:
+          GEOS_THROW( message.str(), InputError );
+          break;
+        case  FieldSpecificationBase::SetErrorMode::warning:
+          message << GEOS_FMT( "As the simulation includes a SurfaceGenerator, the set may be modified later",
+                               fs.getDataContext() );
+          GEOS_WARNING( message.str() );
+          break;
+      }
```

Comment:
- Remove empty case?
- Why `GEOS_FMT()` if no `{}`?
- Possible to take into account SurfaceGenerator presence? (raise a flag from somewhere, or check the data-repo tree?)

```suggestion
      switch( EnumStrings< FieldSpecificationBase::SetErrorMode >::fromString( wrapper.getDefaultValue() ))
      {
        case FieldSpecificationBase::SetErrorMode::error:
          if( isSurfaceGenCase )
            GEOS_THROW( message.str(), InputError );
          break;
        case  FieldSpecificationBase::SetErrorMode::surfaceGeneratorWarning:
          if( isSurfaceGenCase )
            message << "As the simulation includes a SurfaceGenerator, the set may be modified later";
          GEOS_WARNING( message.str() );
          break;
      }
```

Reply:
cannot remove empty case, compil error

Reply:
still lack the if for the throw, isn't it?